nemo/collections/speechlm2/data/duplex_ear_tts_dataset.py

```diff
@@ -0,0 +1,873 @@
+          timestamp in the original recording, preserving the temporal relationship.
+          This is a segment-level alignment only, not word-level alignment.
+    """
+
+    def __init__(
+        self,
+        tokenizer,
+        frame_length: Seconds,
+        source_sample_rate: int,
+        target_sample_rate: int,
+        input_roles: list[str] = None,
+        output_roles: list[str] = None,
+        p_drop_description: float = 0.0,
+        add_text_bos_and_eos_in_each_turn: bool = True,
+        add_audio_prompt: bool = True,
+        audio_prompt_duration: float = 3.0,
+        num_delay_speech_tokens: int = 0,
+    ):
+        self.tokenizer = tokenizer
+        self.frame_length = frame_length
+        self.source_sample_rate = source_sample_rate
+        self.target_sample_rate = target_sample_rate
+        self.input_roles = set(ifnone(input_roles, ["user"]))
+        self.output_roles = set(ifnone(output_roles, ["agent"]))
+        self.p_drop_description = p_drop_description
+        self.add_text_bos_and_eos_in_each_turn = add_text_bos_and_eos_in_each_turn
+        self.add_audio_prompt = add_audio_prompt
+        self.audio_prompt_duration = audio_prompt_duration
+        self.num_delay_speech_tokens = num_delay_speech_tokens
+
+        # compute source and target samples_per_frame
+        self.source_samples_per_frame = int(self.source_sample_rate * self.frame_length)
+        self.target_samples_per_frame = int(self.target_sample_rate * self.frame_length)
+
+        assert tokenizer.bos is not None, "BOS support in the tokenizer is required for S2S models."
+        assert tokenizer.eos is not None, "EOS support in the tokenizer is required for S2S models."
+
+    def __getitem__(self, cuts: CutSet) -> dict:
+        cuts = cuts.transform_text(_strip_timestamps)
+        source_audio, source_audio_lens = collate_audio(cuts.resample(self.source_sample_rate))
+        target_audio, target_audio_lens = collate_audio(
+            cuts.resample(self.target_sample_rate), recording_field="target_audio"
+        )
+        input_text_tokens, target_token_lens = collate_token_channel(
+            cuts,
+            self.tokenizer,
+            self.frame_length,
+            roles=self.output_roles,
+            add_text_bos_and_eos_in_each_turn=self.add_text_bos_and_eos_in_each_turn,
+        )
+        source_tokens, source_token_lens = collate_token_channel(
+            cuts,
+            self.tokenizer,
+            self.frame_length,
+            roles=self.input_roles,
+            add_text_bos_and_eos_in_each_turn=self.add_text_bos_and_eos_in_each_turn,
+        )
+
+        audio_prompt, audio_prompt_lens = get_audio_prompt(
+            cuts, self.target_sample_rate, roles=self.output_roles, recording_field="target_audio"
+        )
+
+        # ensures that input_text_tokens is not longer than its duration
+        input_text_tokens = input_text_tokens[:, : target_token_lens.max()]
```

Comment:
Why is this truncation needed? Because of audio prompt? Shouldn't we increase target_token_lens instead to reflect the full length of the examples? We should avoid data loss at all costs and instead add support to bucketing length estimator to properly support this data format.